File: LEETCODE-PROBLEMS/37 - Sudoku Solver - Hard/SudokuSolver.py
class SudokuSolver(object): 
    # recursion counter   
    def __init__(self):
        self.rec = 0
    
    # prints matrix in table format
    def printer(self, rec, matrix):
        hor = "-------------------------------------"
        print(hor)
        for r in range(len(matrix)):
            l = "| "
            for c in range(len(matrix[r])):
                l += str(matrix[r][c])
                l += " | "
            print(l)
            print(hor)
    
    # validates n for position x, y
    def validate_n(self, matrix, y, x, n):
        for i in range(0, 9):
            if matrix[y][i] == n:
                return False
        for i in range(0, 9):
            if matrix[i][x] == n:
                return False
        x0 = (x // 3) * 3
        y0 = (y // 3) * 3
        for i in range(0, 3):
            for j in range(0, 3):
                if matrix[y0 + i][x0 + j] == n:
                    return False
        return True 
    
    # recursive function for solving matrix
    def solve(self, matrix):
        self.rec += 1
        for y in range(9):
            for x in range(9):
                if matrix[y][x] == 0:
                    for n in range(1, 10):
                        if self.validate_n(matrix, y, x, n):
                            matrix[y][x] = n
                            self.solve(matrix)
                            matrix[y][x] = 0
                    return
        self.printer(5, matrix)
        print("Recursions: ", self.rec)


# --------------- TEST ---------------
solver = SudokuSolver()
matrix = [
    [5, 3, 0, 0, 7, 0, 0, 0, 0],
    [6, 0, 0, 1, 9, 5, 0, 0, 0],
    [0, 9, 8, 0, 0, 0, 0, 6, 0],
    [8, 0, 0, 0, 6, 0, 0, 0, 3],
    [4, 0, 0, 8, 0, 3, 0, 0, 1],
    [7, 0, 0, 0, 2, 0, 0, 0, 6],
    [0, 6, 0, 0, 0, 0, 2, 8, 0],
    [0, 0, 0, 4, 1, 9, 0, 0, 5],
    [0, 0, 0, 0, 8, 0, 0, 7, 9]
]

solver.solve(matrix)

File: LEETCODE-PROBLEMS/37 - Sudoku Solver - Hard/SudokuSolver.java
public class SudokuSolver {
    private int rec;

    private void print(int[][] matrix) {
        String hor = "-------------------------------------";
        System.out.println(hor);
        for (int r = 0; r < matrix.length; r++) {
            String l = "| ";
            for (int c = 0; c < matrix[r].length; c++) {
                l += String.format("%1s | ", matrix[r][c]);
            } System.out.println(l); System.out.println(hor);
        }
    }

    private boolean validate_n(int[][] matrix, int y, int x, int n) {
        for (int i = 0; i < 9; i++) { // Check row
            if (matrix[y][i] == n) { return false; }
        }
        for (int i = 0; i < 9; i++) { // Check column
            if (matrix[i][x] == n) { return false; }
        }
        int x0 = (x / 3) * 3;
        int y0 = (y / 3) * 3;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                if (matrix[y0 + i][x0 + j] == n) { return false; }
            }
        }
        return true;
    }
    
    public void solve(int[][] matrix) {
        rec++;
        for (int y = 0; y < 9; y++) {
            for (int x = 0; x < 9; x++) {
                for (int n = 1; n < 10; n++) {
                    if (validate_n(matrix, y, x, x)) {
                        matrix[y][x] = n;
                        solve(matrix);
                        matrix[y][x] = 0;
                    }
                } return;
            }
        } print(matrix);
    }
}
